src/docker/dockerfileParser.ts

```diff
@@ -0,0 +1,79 @@
+import * as fs from 'fs';
+import { CommandEntry, parse } from 'docker-file-parser';
+
+import { IDockerfile, IDockerParser } from "./parser";
+import { shell, ShellResult } from "../shell";
+
+class RawDockerfile {
+    private readonly commandEntries: CommandEntry[];
+
+    constructor(private readonly dockerfilePath: string) {
+        const dockerData = fs.readFileSync(dockerfilePath, 'utf-8');
+        this.commandEntries = parse(dockerData, { includeComments: false });
+    }
+
+    public getCommandsOfType(...commands: string[]): CommandEntry[] {
+        return this.commandEntries.filter((entry) => {
+            const cmdName = entry.name.toLowerCase();
+            return commands.find((command) => command === cmdName);
+        });
+    }
+
+    public mergeCommandArgsOfType(command: string): string[] {
+        let args = [];
+        this.commandEntries.forEach((entry) => {
+            if (entry.name.toLowerCase() === command) {
+                args = args.concat(entry.args);
+            }
+        });
+        return args;
+    }
+
+    public searchInArgs(regularExpression: RegExp, commands?: string[]): RegExpMatchArray {
+        const commandEntries = (commands ? this.getCommandsOfType(...commands) : this.commandEntries);
+        for (const entry of commandEntries) {
+            const args = Array.isArray(entry.args) ? entry.args : [ String(entry.args) ];
+            for (const arg of args) {
+                const matches = arg.match(regularExpression);
+                if (matches && matches.length) {
+                    return matches;
+                }
+            }
+        }
+
+        return null;

```

Comment:
I'd normally return an empty array rather than a null (which requires special handling).

Reply:
In my understanding, if returning an empty array, the caller still needs use the `array.length` to check it's a valid result. Returning `null` could let the user to boolean expression to check it.

Reply:
But if the caller needs to explicitly check for no matches, don't they need to check the empty array case anyway?  (I know you currently check `matches.length` so that it never returns an empty array, but that's not something I'd trust to remain true as other people maintain the function.)  Either an empty array does 'the right thing' inherently (e.g. a loop scenario) or the caller needs to check for it just in case.

(Or we need a test which asserts if the function ever returns an empty array, so that maintainers are alerted to this detail of the contract.)

Reply:
Agree. thanks for your opinion.